drag-and-drop.md

```diff
@@ -0,0 +1,126 @@
+* Add drag and drop support https://github.com/fyne-io/fyne/issues/142
+* Drag and drop to application icon on macOS https://github.com/fyne-io/fyne/issues/497
+
+Existing interface provide a minimum capabilities to programmatically start a drug operation and get notified when it complets.
+This functionality doesn't provide an interface to communicate between widgets within the application nor send/receive drag and drop with other applications.
+
+## Architecture / API
+
+Existing interface will not change to preserve compatibility with existing code.
+The proposed changes will extend the interface and give user missing capabilities.
+
+The **fyne.CanvasObject** wishing to initiate drag operation must implement **fyne.Draggable** interface. No interface changes.
+
+The **fyne.CanvasObject** wishing to receiver the drop must implement **fyne.DragReceiver** interface.
+
+```
+package fyne
+
+type DragReceiver interface {
+	DragDrop(Draggable)
+}
+```
+
+To pass information to receiver the **fyne.Draggable** can also implement **drag.Info** interface.
+
+```
+package drag
+
+type Info interface {
+	Payload() Payload
+}
+```
+
+To initiate a drag-and-drop outside the Fyne application the object implement **drag.AppDraggable** interface. This interface will allow the object prepare the **drag.Payload**.
+
+```
+package drag
+
+type AppDraggable interface {
+	// StartDrag return true when object ready to start drag-and-drop
+	StartDrag() bool
+	// Payload return Payload to attach to drag-and-drop session
+	Payload() Payload
+}
+```
+
+The initial implementations for **drag.Payload** will include **drag.PayloadString**, **drag.PayloadSliceOfString**. Other types can be added as needed and assuming the platform supports such type.
+
+```
+package drag
+
+type Payload interface {}
+
+type PayloadString string
+type PayloadSliceOfString []string
+
+```
+
+To receive the drag-and-drop from outside the Fyne application the **fyne.CanvasObject** must implement the **fyne.DragReceiver** interface.
+
+The **fyne.Draggable** can implement **drag.Paintable** interface to provide an image that can be used to paint a cursor during drag-and-drop operation.
+
+```
+package drag
+
+type Paintable interface {
+	CursorImage() image.Image
+}
+
+```
+
+## Implementation
+
+### Drag-and-Drop within the Fyne application
+
+The drag and drop within the Fyne application is implemented by finding the component under the mouse pointer and supporting **fyne.DragReceiver** interface in each of the windows of application sorted by z-index. Only the first window with object of **fyne.DragReceiver** will get the call with fyne.Draggable.
+
+The macOS implementation was tested to confirm this design.
+
+This require for each platform to provide a window position, based on its order from front to back among all visible application windows.
+
+The mobile implementation will be concern only with visible window.
+
+### Drop from outside into Fyne application
```

Comment:
I think this section is talking about windows not applications - otherwise it would not discuss canvas objects right? As at the top I think we need to be clear about window-drop onto graphical objects and application drop onto the launching icon?

Reply:
correct, not an icon. It is about dragging a file from desktop to a Tree or List widget of Fyne application.

Reply:
Hopefully we can design this so that a drag action dropped in would not have to behave differently whether it was a Fyne app or not?